pedersen/src/pedersen.rs

```
@@ -0,0 +1,118 @@
+use crate::hash_context::get_standard_pedersen_hash_context;
+use ark_ff::{BigInt, BigInteger, Field, PrimeField};
+use felt::Felt252;
+use num_bigint::BigUint;
+
+pub struct PedersenHash {
+    state: Felt252,
+}
+
+impl PedersenHash {
+    pub fn init_digest_to(digest: Vec<u8>) -> Self {
+        PedersenHash {
+            state: bytes_to_field(&digest),
+        }
+    }
+
+    pub fn get_digest(&self) -> Vec<u8> {
+        self.state.into_bigint().to_bytes_be()
+    }
+
+    pub fn hash(val0: PedersenHash, val1: PedersenHash) -> PedersenHash {
+        let cntx = get_standard_pedersen_hash_context();
+        let res = cntx.hash_elements(val0.state, val1.state);
+        PedersenHash { state: res }
+    }
+
+    pub fn hash_bytes_with_length(bytes: Vec<u8>) -> PedersenHash {
+        let mut state = Felt252::ZERO;
+        let cntx = get_standard_pedersen_hash_context();
+        let modulus = BigUint::from_bytes_be(&Felt252::MODULUS.to_bytes_be());
+
+        let mut bytes_to_hash = bytes.len();
+        let mut offset = 0;
+
+        while bytes_to_hash >= 32 {
+            let word = BigUint::from_bytes_be(&bytes[offset..offset + 32]);
+            let q = word.clone() / modulus.clone();
+            let r = word % modulus.clone();
+            let value = bytes_to_field(&r.to_bytes_be());
+            assert!(q < BigUint::from(1000u64), "Unexpectedly large shift.");
+            let shift = bytes_to_field(&q.to_bytes_be());
+
+            state = cntx.hash_elements(state, value) + shift;
+
+            offset += 32;
+            bytes_to_hash -= 32;
+        }
+
+        assert!(
```

nitpick: assert_eq

---

Move the assert to the top, to make it clear that it is a precondition of the function body: fail early if the precondition is not satisfied.